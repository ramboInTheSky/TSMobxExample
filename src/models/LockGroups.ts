export class LockGroup{
    constructor(id?: string, name?: string, type?: string){
        this.id = id || ''
        this.name = name || ''
        this.type = type || 'Room'
    }
    id: string = ''
    name: string = ''
    type: string = ''
    notes: string = ''
    selected: boolean = false
}

export class CreateLockGroup{
    name: string = ''
    type: string = ''
    notes: string = ''
}

export interface LockGroupResponse {
	count: number
	filters: Array<string>
	items: Array<LockGroup>
}
