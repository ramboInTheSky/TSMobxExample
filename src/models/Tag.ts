export class Tag{
    tagNumber: string | null = null

    constructor(tag?: string){
        this.tagNumber = tag || null
    }
}