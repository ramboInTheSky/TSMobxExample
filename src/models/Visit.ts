import { LockGroup } from './LockGroups'
export class Visit {
    id: string = ''
    lockGroups?: Array<LockGroup | string>  = []
    fromDate: string = ''
    toDate: string = ''
    tags: Array<string> = []
    archetypeIds: Array<string> = []
}

export class CreateVisit {
    lockGroups?: Array<LockGroup> = []
    fromDate: string = ''
    toDate: string = ''
    tags: Array<string> = []
    archetypeIds: Array<string> = []
}

export interface Archetype {
    id: string
    name: string
    primary: boolean
    lockGroupNames: Array<string>
}

export class Archetypes {
    primary: Array<Archetype> = []
    secondary: Array<Archetype> = []
}