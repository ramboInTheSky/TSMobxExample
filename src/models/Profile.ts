import { Visit } from './Visit'

export type ProfileType = 'visitor' | 'member'

export class Profile {
	id?: string = undefined
	fullName?: string = undefined
	firstName: string = ''
	lastName: string = ''
	company?: string = undefined
	mobileNumber?: string = ''
	emailAddress?: string = ''
	notes?: string = undefined
	created?: string = undefined
	type?: ProfileType = undefined
	visits?: Array<Visit> = undefined
}

export class CreateProfile {
	firstName: string = ''
	lastName: string = ''
	company?: string = undefined
	mobileNumber?: string = ''
	emailAddress?: string = ''
	notes?: string = undefined
	type?: ProfileType = undefined
}

export class EditProfile {
	id?: string = undefined
	firstName: string = ''
	lastName: string = ''
	company?: string = undefined
	mobileNumber?: string = ''
	emailAddress?: string = ''
	notes?: string = undefined
	created?: string = undefined
	type?: ProfileType = undefined
}

export interface ProfileResponse {
	count: number
	filters: Array<string>
	items: Array<Profile>
}
