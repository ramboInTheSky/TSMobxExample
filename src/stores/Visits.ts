import { observable, action, computed } from 'mobx'
import axios from 'axios'
import { Store } from './APIStore'
import { Visit, CreateVisit, Archetype } from '../models/Visit'
import urls from '../constants/urls'
import { Tag } from '../models/Tag';
import { toSnakeCase, toCamelCase, getErrorCode } from '../utils';
import { error } from 'util';
import { LockGroup } from '../models/LockGroups';

const sortingFn = (a: Visit, b: Visit) => {
	const first = new Date(a.toDate)
	const last = new Date(b.toDate)
	if (first > last) return -1
	if (last > first) return 1
	return 0
}

export class VisitStore extends Store<Visit> {
	@observable protected archetypesObj: {
		primary: Array<Archetype>,
		secondary: Array<Archetype>
	} = {
			primary: [],
			secondary: []
		}

	@observable protected edit: boolean = false
	@observable protected roomLoading: {[key: string]:boolean} = {}
	@observable protected internalValidationError?: string = undefined
	@observable protected erroredTag?: string = undefined

	public get editing() {
		return this.edit
	}

	public get isValidationError() {
		return !!this.validationError
	}

	public isRoomLoading(id: string) {
		return !!this.roomLoading[id]
	}

	@action
	clearErrors = () => {
		this.internalValidationError = undefined
		this.erroredTag = undefined
	}

	public get validationError() {
		return this.internalValidationError
	}

	@action
	toggleEditing(editing?: boolean) {
		if (editing !== undefined) {
			this.edit = editing
		} else this.edit = !this.edit
	}


	public get archetypes() {
		return this.archetypesObj
	}

	constructor() {
		super(urls.visits)
	}

	async getArchetypes() {
		const data = await super.getList(urls.archetypes, undefined)
		const primary = data.items.filter((item: Archetype) => item.primary)
		const secondary = data.items.filter((item: Archetype) => !item.primary)
		this.archetypesObj = { primary, secondary }
	}

	async save(item: Visit, profileId: string) {
		let timeout: any = undefined
		try {
			let tries = 0
			const maxTries = 30
			const pollingTime = 3000
			const url = `${urls.profiles}/${profileId}/visit`
			const payload = Object.assign(new CreateVisit(), item)
			this.detail = {}
			const data = await super.save(payload, undefined, url)
			this.loading = true
			this.edit = false
			this.detail = data
			const self = this
			try {
				let pollingResult: any = {}
				timeout = setInterval(async () => {
					try {
						if (tries < maxTries) {
							pollingResult = await axios.get(`${url}/${self.detail.id}/status`)
							tries++
							if (pollingResult && pollingResult.data.finished) {
								if (!pollingResult.data.error) {
									self.apiEnded()
								}
								else {
									self.apiEnded(pollingResult, pollingResult.data.error)
								}
								clearInterval(timeout)
							}
						}
						else {
							throw new Error('maximum number of polls exceeded')
						}
					}
					catch (pollingError) {
						this.apiEnded(pollingError)
						clearInterval(timeout)
					}
				}, pollingTime)
			}
			catch (err) {
				this.apiEnded(err, 'TEMP MESSAGE - Something has gone wrong with Salto!')
			}
		}
		catch (failsafe) {
			clearInterval(timeout)
		}
	}

	async addTag(tag: Tag, profileId: string, visitId: string) {
		const url = `${urls.profiles}/${profileId}/visit/${visitId}/tag`
		let res: any
		try {
			this.loading = true
			res = await axios.post(url, toSnakeCase(tag))
			this.apiEnded()
			const data = toCamelCase(res.data)
			if (data.success) {
				this.detail.tags = this.detail.tags || []
				this.detail.tags.push(data.tagNumber)
			}
			else {
				this.apiEnded()
				this.erroredTag = data.tagNumber
				this.internalValidationError = getErrorCode(data)
			}
			return data.success
		} catch (error) {
			this.apiEnded()
			let data = {}
			if (error.response) {
				data = toCamelCase(error.response.data)
			}
			this.erroredTag = tag.tagNumber!
			this.internalValidationError = getErrorCode(data)
			return {}
		}
	}

	async addAccess(lockGroup: LockGroup, profileId: string, visitId: string) {
		const url = `${urls.profiles}/${profileId}/visit/${visitId}/lockgroup`
		let res: any
		try {
			this.roomLoading[lockGroup.id] = true
			res = await axios.post(url, {
				"lock_group_id": lockGroup.id
			})
			const data = toCamelCase(res.data)
			if (data.lockGroupId) {
				this.detail.lockGroups = this.detail.lockGroups || []
				this.detail.lockGroups.push(lockGroup as LockGroup)
				return true
			}
			this.roomLoading[lockGroup.id] = false
			return false
		} catch (error) {
			if (error.response) {
				this.apiEnded(toCamelCase(error.response.data))
			}
			else {
				this.apiEnded(error)
			}
			this.roomLoading[lockGroup.id] = false
			return {}
		}
	}
}