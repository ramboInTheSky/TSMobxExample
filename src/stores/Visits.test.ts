import React from 'react'
import axios from 'axios'

import { VisitStore } from './Visits'
import { CreateVisit, Visit } from '../models/Visit';

jest.mock('axios', () => ({
    get: jest.fn(
        (url: string) =>
            url === 'fail' ? Promise.reject({ status: 500 }) : Promise.resolve({ status: 200, data: {} }),
    ),
    post: jest.fn(
        (url: string) =>
            url === 'fail' ? Promise.reject({ status: 500 }) : Promise.resolve({ status: 200, data: {} }),
    ),
}))

describe('<VisitStore />', () => {
    it('renders correctly', () => {
        const store = new VisitStore()
        expect(store).toMatchSnapshot()
    })
    it('sets isError if API issues', async () => {
        const store = new VisitStore()
        await store.getList()
        setTimeout(() => expect(store.isError).toBe(true))
    })
    it('returns data correctly', async () => {
        const store = new VisitStore()
        await store.getList()
        setTimeout(() => expect(store.isError).toBe(false))
    })
    it('saves correctly and polls for status', async () => {
        const store = new VisitStore()
        const visit = Object.assign(new Visit(), { "lock_groups": [], "from_date": "2019-05-11T16:40:00.000Z", "to_date": null, "tags": [], "archetype_ids": ["9108859f-c93a-48dd-8e49-8725390d063d"] })
        await store.save(visit, 'profile-id')
        expect(axios.post).toHaveBeenCalledWith(
            "https://csi-dev.thecollective.com/ac-api/profile/profile-id/visit",
            {
                "archetype_ids": ["9108859f-c93a-48dd-8e49-8725390d063d"],
                "from_date": "2019-05-11T16:40:00.000Z",
                "id": null,
                "lock_groups": [],
                "tags": [],
                "to_date": null
            })

        expect(axios.get).toHaveBeenCalled()
        //TODO test polling better
    })

})
