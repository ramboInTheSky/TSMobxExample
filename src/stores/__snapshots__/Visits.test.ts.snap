// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VisitStore /> renders correctly 1`] = `
VisitStore {
  "apiEnded": [Function],
  "fresh": false,
  "isFresh": [Function],
  "url": "https://csi-dev.thecollective.com/ac-api/visit",
}
`;
